lib/vquic/curl_quiche.c

```diff
@@ -216,6 +216,15 @@ static CURLcode quic_ssl_setup(struct Curl_cfilter *cf, struct Curl_easy *data)
     }
   }
 
+  {
+    struct connectdata *conn = data->conn;
+    const char *curves = conn->ssl_config.curves;
+    if(curves && !SSL_CTX_set1_curves_list(ctx->sslctx, curves)) {
+      failf(data, "failed setting curves list for QUIC: '%s'", curves);
+      return CURLE_SSL_CIPHER;
+    }
+  }
```

Comment:
This has to be set higher up before the SSL object is created from the SSL_CTX. In other words above SSL_new. There is no similar code for ciphers here, is that intentional? If both these backends need similar CTX init or SSL init (eg same SNI code is now in both) maybe it would be better to have a common function rather than duplicate the code

Also I don't have quiche but looking at this it does not appear to free SSL_CTX and SSL in cf_quiche_ctx_clea? not sure if that's a bug or not but ngtcp2 clear there's similar code that does that.

Reply:
> This has to be set higher up before the SSL object is created from the SSL_CTX. In other words above SSL_new

That sounds wrong. This functions uses the sslctx as argument so it needs to be done after that is created, which is done within this same function.

> There is no similar code for ciphers here, is that intentional?

Yes, this is using BoringSSL and I don't know how those are set other than this.

> looking at this it does not appear to free SSL_CTX and SSL in cf_quiche_ctx_clear

True, but that seems unrelated to this PR.

Reply:
> > This has to be set higher up before the SSL object is created from the SSL_CTX. In other words above SSL_new
> 
> That sounds wrong. This function uses the sslctx as argument so it needs to be done after that is created, which is done within this same function.

When an SSL object is created it points to the CTX used to create it, not a copy. In other words:

~~~c
ctx = SSL_CTX_new(TLS_client_method());
ssl = SSL_new(ctx);
assert(ctx == SSL_get_SSL_CTX(ssl));
~~~

The OpenSSL documentation [says](https://www.openssl.org/docs/man3.0/man3/SSL_CTX_new.html):

_"**An SSL_CTX object should not be changed after it is used to create any SSL objects** or from multiple threads concurrently, since the implementation does not provide serialization of access for these cases."_

You are doing this:
~~~c
ctx->ssl = SSL_new(ctx->sslctx);
...
SSL_CTX_set1_curves_list(ctx->sslctx, curves);
~~~

It may appear to work but it's not guaranteed. I am saying do this instead:
~~~c
SSL_CTX_set1_curves_list(ctx->sslctx, curves);
...
ctx->ssl = SSL_new(ctx->sslctx);
~~~

Reply:
aah now I see what you mean. Pushed an update